web-services/common-util/src/test/java/datawave/security/authorization/DatawavePrincipalTest.java

```diff
@@ -0,0 +1,189 @@
+        Assert.assertEquals(server1SubjectDn, dp.getPrimaryUser().getDn().subjectDN());
+        
+        // these tests are for case where a UserType.USER appears anywhere in the proxiedUsers collection
+        dp = new DatawavePrincipal(Lists.newArrayList(user, server1, server2, server3));
+        Assert.assertEquals(userSubjectDn, dp.getPrimaryUser().getDn().subjectDN());
+        
+        dp = new DatawavePrincipal(Lists.newArrayList(server1, user, server2, server3));
+        Assert.assertEquals(userSubjectDn, dp.getPrimaryUser().getDn().subjectDN());
+        
+        dp = new DatawavePrincipal(Lists.newArrayList(server1, server2, user, server3));
+        Assert.assertEquals(userSubjectDn, dp.getPrimaryUser().getDn().subjectDN());
+        
+        dp = new DatawavePrincipal(Lists.newArrayList(server1, server2, server3, user));
+        Assert.assertEquals(userSubjectDn, dp.getPrimaryUser().getDn().subjectDN());
+    }
+    
+    @Test
+    public void GetProxyServersTest() {
+        // direct call from finalServer
+        DatawavePrincipal dp = new DatawavePrincipal(Lists.newArrayList(finalServer));
+        Assert.assertEquals(null, dp.getProxyServers());
+        
+        // direct call from user
+        dp = new DatawavePrincipal(Lists.newArrayList(user));
+        Assert.assertEquals(null, dp.getProxyServers());
+        
+        // call from finalServer proxying initial caller server1
+        dp = new DatawavePrincipal(Lists.newArrayList(finalServer, server1));
+        Assert.assertEquals(Arrays.asList(finalConnectionServerSubjectDn), dp.getProxyServers());
+        
+        // call from finalServer proxying initial caller server1 through server2
+        dp = new DatawavePrincipal(Lists.newArrayList(finalServer, server1, server2));
+        Assert.assertEquals(Arrays.asList(server2SubjectDn, finalConnectionServerSubjectDn), dp.getProxyServers());
+        
+        // call from finalServer proxying initial caller server1 through server2 and server3
+        dp = new DatawavePrincipal(Lists.newArrayList(finalServer, server1, server2, server3));
+        Assert.assertEquals(Arrays.asList(server2SubjectDn, server3SubjectDn, finalConnectionServerSubjectDn), dp.getProxyServers());
+        
+        // these tests are for cases where a UserType.USER appears anywhere in the proxiedUsers collection
+        
+        // this first case would be very odd -- call from user proxying initial caller server1 through server2 through server3
+        dp = new DatawavePrincipal(Lists.newArrayList(user, server1, server2, server3));
+        Assert.assertEquals(Arrays.asList(server1SubjectDn, server2SubjectDn, server3SubjectDn), dp.getProxyServers());
+        
+        dp = new DatawavePrincipal(Lists.newArrayList(server1, user, server2, server3));
+        Assert.assertEquals(Arrays.asList(server2SubjectDn, server3SubjectDn, server1SubjectDn), dp.getProxyServers());
```

Comment:
Just to make sure I understand this correctly, this case is saying that the call order was:

server2 submits the request to server3 (on behalf of "user")
server3 submits the request to server1
server1 submits the request to datawave

Or is it this?
server3 submits the request to server1 (on behalf of "user")
server1 submits the request to server2
server2 submits the request to datawave

Or something else?

Reply:
Ok, cool.  thanks for clearing that up.  i may forget in a month, but i can always refer back to this pr comment, ha.